content/posts/canvas-renderer-matplotlib-in-pyodide.md

```diff
@@ -0,0 +1,193 @@
+
+To make this possible — the function [putImageData()](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/putImageData) is used for off-screen rendering. This is used to put an ImageData object into an off-screen canvas. Once that is done, the on-screen canvas can now use the [drawImage()](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage) function — to copy the contents of the off-screen canvas and render that stuff to the screen.
+
+While the long process of rendering to off-screen and then transferring its data to on-screen using *drawImage()* seems like overkill, it helps us avoid the 2 issues regarding transformations and transparency we discussed above.
+
+### **Results**
+
+With the above major issues fixed, the new backend is ready. Here are some sample plots (taken from the Matplotlib gallery) that are rendered using the new canvas based backend.
+
+![results](/posts/2/results.png)
+
+### **How Fast is it? and Potential Optimizations**
+
+Using the Canvas API to draw the plots live introduces a performance penalty. We assume this is due to iterating over all of the points on a curve in Python, since loops in Python are known to be slow. Further, loading fonts asynchronously over the network can take some time. However, Rendering Images is as fast as before. Below are the benchmarks for the new Canvas based renderer compared to the default Agg renderer for 5 sample plots in both firefox and chrome.
+
+![metrics](/posts/2/metrics.png)
+
+In essence, the new renderer is about 1.25 to 1.5x slower but that’s the price one pays for reducing the size.
+
+To overcome the above slowdown due to the rendering loop in Python, one solution is to rewrite this time-critical piece in JavaScript. This can be possible using the Function construct in JavaScript. The Function construct would essentially allow us to create JavaScript functions inside a Python module capable of accepting and working with pythonic arguments.
+
+![snippet](/posts/2/snippet.png)
+
+The above snippet defines a JavaScript function *_path_helper* which accepts Pythonic arguments and also supports calling of pythonic object’s methods inside it -- *path.iter_segments()*. The work related to this was attempted here: https://github.com/pyodide/pyodide/pull/510
+
+### **Conclusion**
+
+It’s been really gratifying to see all of the cool things that have been created with Pyodide in the few years since its public launch.  However, there’s still lots to do to turn this proof-of-concept into a professional tool for everyday data science work. If you’re interested in helping us build that future, come find us on Gitter, Github and our mailing list.
+
+---
+
```

Comment:
Do you want to add an "About the author" section, as was done in https://github.com/pyodide/pyodide-blog/blame/master/content/posts/react-in-python-with-pyodide.md#L435

Reply:
added, thanks